state/cluster/badger/mutator.go

```diff
@@ -36,202 +37,286 @@ func NewMutableState(state *State, tracer module.Tracer, headers storage.Headers
+	}
+	setupSpan.End()
 
-		// get the latest finalized cluster block and latest finalized consensus height
-		var finalizedClusterBlock flow.Header
-		err := procedure.RetrieveLatestFinalizedClusterHeader(chainID, &finalizedClusterBlock)(tx)
-		if err != nil {
-			return fmt.Errorf("could not retrieve finalized cluster head: %w", err)
-		}
-		var finalizedConsensusHeight uint64
-		err = operation.RetrieveFinalizedHeight(&finalizedConsensusHeight)(tx)
-		if err != nil {
-			return fmt.Errorf("could not retrieve finalized height on consensus chain: %w", err)
-		}
+	checkAncestrySpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckAncestry)
+	// ensure that the extending block connects to the finalized state, we
+	// do this by tracing back until we see a parent block that is the
+	// latest finalized block, or reach height below the finalized boundary
 
-		// get the header of the parent of the new block
-		parent, err := m.headers.ByBlockID(header.ParentID)
-		if err != nil {
-			return fmt.Errorf("could not retrieve latest finalized header: %w", err)
-		}
+	// start with the extending block's parent
+	parentID := header.ParentID
+	for parentID != extendCtx.finalizedClusterBlock.ID() {
 
-		// extending block must have correct parent view
-		if header.ParentView != parent.View {
-			return state.NewInvalidExtensionErrorf("candidate build with inconsistent parent view (candidate: %d, parent %d)",
-				header.ParentView, parent.View)
+		// get the parent of current block
+		ancestor, err := m.headers.ByBlockID(parentID)
+		if err != nil {
+			return fmt.Errorf("could not get parent (%x): %w", block.Header.ParentID, err)
 		}
 
-		// the extending block must increase height by 1 from parent
-		if header.Height != parent.Height+1 {
-			return state.NewInvalidExtensionErrorf("extending block height (%d) must be parent height + 1 (%d)",
-				block.Header.Height, parent.Height)
+		// if its height is below current boundary, the block does not connect
+		// to the finalized protocol state and would break database consistency
+		if ancestor.Height < extendCtx.finalizedClusterBlock.Height {
+			return state.NewOutdatedExtensionErrorf("block doesn't connect to finalized state. ancestor.Height (%d), final.Height (%d)",
+				ancestor.Height, extendCtx.finalizedClusterBlock.Height)
 		}
 
-		// ensure that the extending block connects to the finalized state, we
-		// do this by tracing back until we see a parent block that is the
-		// latest finalized block, or reach height below the finalized boundary
-
-		setupSpan.End()
-		checkAnsSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckAncestry)
+		parentID = ancestor.ParentID
+	}
+	checkAncestrySpan.End()
 
-		// start with the extending block's parent
-		parentID := header.ParentID
-		for parentID != finalizedClusterBlock.ID() {
+	checkRefBlockSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckReferenceBlock)
```

Comment:
also here concerned about span potentially not being closed

Reply:
https://github.com/onflow/flow-go/pull/4148/commits/5fdbee811aed60889e941a9fd0ab44dd1a576c4c